### DISCUSSION TITLE
Plugin creation: How do you loop over all selected files

### DISCUSSION THREAD

**Discussion by thepsauce** [sxyazi/yazi#1117] [1. Q&A] [answered]
*I have answered my own question by looking through the [github page](https://github.com/sxyazi/yazi/blob/main/yazi-plugin/preset/plugins) and the [website](https://yazi-rs.github.io/docs/plugins/). The solution is at the bottom..*

Hello, I'm trying to make a plugin in yazi using lua. I want to get input from the user and loop over all selected files (marked with space or any other method). This was my attempt:
```lua
local M = {}

function M:entry()
    local value, event = ya.input {
	    title = "Tag file",
	    position = { "center", w = 50 },
    }
    if event == 1 then
        for idx, url in pairs(cx.yanked) do
            ya.notify {
                title = "you",
                content = url.name,
                timeout = 2,
            }
        end
    end
end

return M
```
But it can't run the loop after `if event == 1 then`, it gives this error:
```
Micro plugin failed:
runtime error: [string "/usr/src/debug/yazi/yazi-0.2.5/yazi-plugin/sr..."]:9: attempt to index a nil value (global 'cx')
stack traceback:
	[string "/usr/src/debug/yazi/yazi-0.2.5/yazi-plugin/sr..."]:9: in function <[string "/usr/src/debug/yazi/yazi-0.2.5/yazi-plugin/sr..."]:3>
```

What I actually want to do is for each selected file create a symlink using `ln -s <file> <name>` when a key is pressed.

Edit: I got one step further!

```lua
local M = {}

local tag_files = ya.sync(function(state)
	local h = cx.active.current.hovered
    if h then
        ya.notify {
            title = "tagged",
            content = tostring(h.url),
            timeout = 2,
        }
    end
    for idx, url in pairs(cx.yanked) do
        ya.notify {
            title = "tagged",
            content = tostring(url),
            timeout = 2,
        }
    end
end)

function M:entry()
    local value, event = ya.input {
	    title = "Tag file",
	    position = { "center", w = 50 },
    }
    if event == 1 then
        tag_files()
    end
end

return M
```
This now works for the currently hovered file but not the selected files, but now I noticed that `cx.yanked` is probably not the selected files but actually the yanked files.. so my question remains.

Edit2: I figured it out after viewing [this](https://github.com/sxyazi/yazi/blob/main/yazi-plugin/preset/plugins/folder.lua), `Folder.CURRENT` just worked instead of `PREVIEW`.
```lua
local folder = Folder:by_kind(Folder.CURRENT)
if not folder then
    return
end

for i, f in ipairs(folder.window) do
    local marker = File:marker(f)
    if marker ~= 0 then
        ya.notify {
            title = "tagged",
            content = tostring(f.url),
            timeout = 2,
        }
    end
end
```

**Comment by sxyazi** [comment, marked as answer]
It is not recommended to use `ipairs(folder.window)` since it needs to iterate over all files, which is inefficient and can only get the visible files in the window.

- If you want to access the selected file(s), use `cx.active.selected`
- If you want to access the yanked file(s), use `cx.yanked`

They both need to be used in a [sync context](https://yazi-rs.github.io/docs/plugins/overview#sync-context).

**Reply by thepsauce** [reply]
Great, thank you! This worked:
```lua
for i, f in pairs(cx.active.selected) do
    tag_file(f:name(), tags)
end
```
